py/embedded-server/deephaven_server/__init__.py

```diff
@@ -14,3 +19,55 @@
 # distribution name from the package name.
 # https://docs.python.org/3/library/importlib.metadata.html#package-distributions
 __version__ = importlib.metadata.version('deephaven-server')
+
+
+def main():
+    """
+    Main entry point for the Deephaven embedded server application.
+    Accepts a command to start the application and the arguments to use.
+    Defaults to `start`, which starts the server.
+    """
+    parser = argparse.ArgumentParser(description="Start the Deephaven embedded server")
+    parser.add_argument("cmd", nargs="?", default="start", choices=["start"], help="Deephaven command to run. Defaults to start.")
+    parser.add_argument("-p", "--port", default=8080, type=int, help="The port to bind to. Defaults to 8080.")
+    parser.add_argument("--host", default="localhost", help="The host to bind to. Defaults to localhost.")
+    parser.add_argument("--key", default=None, help="The key to use. Uses a randomly generated key if not specified.")
+    parser.add_argument("--jvm-args", default=None, help="The JVM arguments to use.")
+    parser.add_argument("--dh-args", default=None, help="The Deephaven arguments to use.")
+    args = parser.parse_args()
+
+    if args.cmd == "start":
+        _start(args)
+    else:
+        parser.print_help()
+        sys.exit(1)
+
+def _start(args):
+    """
+    Start the Deephaven server with the given arguments.
+
+    Args:
+        args: The arguments parsed using argparse to use to start the server.
+    """
+    if args.key is None:
+        args.key = secrets.token_urlsafe(32)
+
+    jvm_args = [f"-Dauthentication.psk={args.key}"]
```

Comment:
I tried to run locally, and these default jvm_args / host / port break my setup.
![Screenshot from 2024-03-21 19-08-22](https://github.com/deephaven/deephaven-core/assets/6764691/17d340fc-c57b-43ac-9687-14c778c73ea7)


Specifically, I run with:

```
$ cat ~/.config/deephaven/deephaven.prop                         
includefiles=dh-defaults.prop

http.port=8443
http.http1=false

ssl.identity.type=privatekey
ssl.identity.certChainPath=/home/devin/.config/deephaven/dev-certs/server.chain.crt
ssl.identity.privateKeyPath=/home/devin/.config/deephaven/dev-certs/server.key

AuthHandlers=io.deephaven.auth.AnonymousAuthenticationHandler
```

IMO, I think we either need to respect the default (or explicitly pointed to) config file (we can set up a proper way to query the config file in a proper way [ie, _not_ just reading it] so the client can get out the host, port); or, we need to say that deephaven-server is special and has its own config application namespace. That said, even if it _does_ have its own application namespace, there is still room for configuration files at that location that aren't compatible with this setup...

Reply:
Yes that's a good point. I could get rid of the defaults and then it shouldn't conflict. However, the biggest issue I see is with the key. If I do not know the pre-shared key, I can't print the URL to which to open the UI for. I could still open the UI, but you'll be prompted to enter the key... which you won't know if you didn't specify it in your configuration file, which as a new user ad-hoc testing it out, is unlikely to have done.

So what's the best way to resolve it such that I can get the key that is configured? And/or get the key the server randomly generates?

Reply:
We could probably hook up a generic way to get a URL from the server in process; this would allow it to _potentially_ work with other authorization methods if they chose to implement it. At a minimum, we'd implement it for anonymous and PSK. And it would also handle https cases. In my configuration above, the server would return `https://localhost:8443/ide/` since I'm running TLS + 8443 + anonymous.

Reply:
I've changed it so it doesn't default a port, it allows the server startup to initialize it/read it from config.
The key I'm still generating, though also added a flag to run in `anonymous` mode... so in your case, it _should_ work for you if you run `deephaven --anonymous`.

> We could probably hook up a generic way to get a URL from the server in process

Sure, but that doesn't exist right now. Would you rather wait for that to get wired up and then change this to match? Or merge that down the road? Or change this PR in some other way?